server/src/main/java/org/opensearch/search/pipeline/ProcessorExecutionDetail.java

```diff
@@ -0,0 +1,246 @@
+        } else {
+            builder.field(fieldName, data);
+        }
+    }
+
+    private void writeItemToXContent(XContentBuilder builder, Object item, Params params) throws IOException {
+        if (item instanceof ToXContentObject) {
+            ((ToXContentObject) item).toXContent(builder, params);
+        } else {
+            builder.value(item);
+        }
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        ProcessorExecutionDetail that = (ProcessorExecutionDetail) o;
+        return durationMillis == that.durationMillis
+            && Objects.equals(processorName, that.processorName)
+            && Objects.equals(inputData, that.inputData)
+            && Objects.equals(outputData, that.outputData);
+    }
+
+    public static ProcessorExecutionDetail fromXContent(XContentParser parser) throws IOException {
+        String processorName = null;
+        long durationMillis = 0;
+        Object inputData = null;
+        Object outputData = null;
+        if (parser.currentToken() != XContentParser.Token.START_OBJECT) {
+            parser.nextToken();
+            ensureExpectedToken(XContentParser.Token.START_OBJECT, parser.currentToken(), parser);
+        }
+        while (parser.nextToken() != XContentParser.Token.END_OBJECT) {
+            String fieldName = parser.currentName();
+            parser.nextToken();
+
+            if (PROCESSOR_NAME_FIELD.match(fieldName, parser.getDeprecationHandler())) {
+                processorName = parser.text();
+            } else if (DURATION_MILLIS_FIELD.match(fieldName, parser.getDeprecationHandler())) {
+                durationMillis = parser.longValue();
+            } else if (INPUT_DATA_FIELD.match(fieldName, parser.getDeprecationHandler())) {
+                inputData = parseFieldFromXContent(parser);
+            } else if (OUTPUT_DATA_FIELD.match(fieldName, parser.getDeprecationHandler())) {
+                outputData = parseFieldFromXContent(parser);
+            } else {
+                parser.skipChildren();
+            }
+        }
+
+        if (processorName == null) {
+            throw new IllegalArgumentException("Processor name is required");
+        }
+
+        return new ProcessorExecutionDetail(processorName, durationMillis, inputData, outputData);
+    }
+
+    private static Object parseFieldFromXContent(XContentParser parser) throws IOException {
+        XContentParser.Token token = parser.currentToken();
+        if (token == XContentParser.Token.VALUE_NULL) {
+            return null;
+        } else if (token == XContentParser.Token.START_ARRAY) {
+            return parseArrayFromXContent(parser);
+        } else if (token == XContentParser.Token.START_OBJECT) {
+            return parser.map();
+        } else {
+            return parser.textOrNull();
+        }
+    }
```

Comment:
I would be surprised if this utility method (and the related array one) doesn't already exist somewhere ([like here?](https://github.com/opensearch-project/OpenSearch/blob/main/libs/x-content/src/main/java/org/opensearch/common/xcontent/XContentUtils.java)).  If it doesn't, perhaps this code should be in such a util class?